EGW: Delete Stale Conntrack Entries

tssurya opened the pull request:
OVNK parts work, ARP-ing works, conntrack deletion works -> tested on KIND and user lab
However in order for this PR to work completely we will need [an upstream PR to netlink library for label matches.](https://github.com/vishvananda/netlink/pull/784) to merge first and then I can remove my 2nd commit.

Logic:

1. annotate the namespace each time an exgw is added/deleted with list of ips
2. add new informer for namespace on node side checking only if gw ip annotation changed OR routing-externalgws annotation changed
3. ovnkube node on namespace change, iterates through all the ips and initiates an arp request via ovnk and collects the MACs
4. once all the responses come back, we have all the known macs
5. we search for ct entries for any pod ip belonging to the namespace, if ct_label is loaded with a mac not in our list we flush it
6. we run the above in a goroutine as well set which will run every 5mins looping through all relevant namespaces.

Brainstormed-By: @trozet , @dcbw 

NOTE for reviewers: Added/Updated unit tests, and two new e2e test's in `external_gateways.go` file for the namespace annotation case and external gateway annotation case.

tssurya commented:
/retest-failed

dcbw commented:
@tssurya looks like the Build-PR failure is a variant of https://github.com/ovn-org/ovn-kubernetes/pull/3062 but in pods_test.go? I think that any time we rely on libovsdb's reconnect behavior to inject failures into our testcases, we should be waiting longer like 3062...

tssurya commented:
> @tssurya looks like the Build-PR failure is a variant of #3062 but in pods_test.go? I think that any time we rely on libovsdb's reconnect behavior to inject failures into our testcases, we should be waiting longer like 3062...
> 
> To be clear, not introduced in your PR, I think it's a flake.

yes! good point. Let me bookmark the run https://github.com/ovn-org/ovn-kubernetes/runs/7221469923?check_suite_focus=true#step:6:15930 for future analysis. Yea I think we do need to fix those tests

TY for the retest :D

dcbw commented on go-controller/pkg/ovn/namespace.go:
```diff
@@ -20,11 +20,13 @@ import (
+	routingNamespaceAnnotation       = "k8s.ovn.org/routing-namespaces"
+	routingNetworkAnnotation         = "k8s.ovn.org/routing-network"
+	bfdAnnotation                    = "k8s.ovn.org/bfd-enabled"
+	externalGatewayPodIPsAnnotations = "k8s.ovn.org/external-gw-podIPs"
```
@tssurya let's keep these all lowercase...

dcbw replied:
Arguably the other namespace annotations should go there too, but we can do that later.

tssurya replied:
yeah let me move all these into types.go where they belong globally?

dcbw commented on go-controller/pkg/ovn/egressgw.go:
```diff
@@ -165,7 +181,6 @@ func (oc *Controller) addPodExternalGW(pod *kapi.Pod) error {
 		klog.Warningf("No valid gateway IPs found for requested external gateway pod: %s", pod.Name)
 		return nil
 	}
-
```
@tssurya any reason to remove the space, or just left-over change from a previous version of the patch?

dcbw commented on go-controller/pkg/ovn/egressgw.go:
```diff
@@ -178,11 +193,11 @@ func (oc *Controller) addPodExternalGW(pod *kapi.Pod) error {
 func (oc *Controller) addPodExternalGWForNamespace(namespace string, pod *kapi.Pod, egress gatewayInfo) error {
 	var gws string
-	for _, ip := range egress.gws {
+	for _, ip := range egress.gws.UnsortedList() {
```
@tssurya since this is now a string list, we can use `gws := strings.Join(egress.gws.UnsortedList(), ",")` maybe?

dcbw commented on go-controller/pkg/ovn/namespace.go:
```diff
@@ -61,8 +63,7 @@ func (oc *Controller) getRoutingExternalGWs(nsInfo *namespaceInfo) *gatewayInfo
 	res.bfdEnabled = nsInfo.routingExternalGWs.bfdEnabled
-	res.gws = make([]net.IP, len(nsInfo.routingExternalGWs.gws))
-	copy(res.gws, nsInfo.routingExternalGWs.gws)
+	res.gws = nsInfo.routingExternalGWs.gws
```
@tssurya won't this pass back the same sets.String struct/object and open up to concurrency issues? I think that's what the copy() was for IIRC.

Should we do `res.gws = nsInfo.routingExternalGWs.gws.Difference(NewString())` or something to return a new set with the same data?

tssurya replied:
`sets.NewString(nsInfo.routingExternalGWs.gws.UnsortedList()...)` went with that, since it seems to give a new set with the given items: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#NewString

dcbw commented on go-controller/pkg/node/node.go:
```diff
@@ -31,6 +31,11 @@ import (
+const (
+	// Annotations used by multiple external gateways feature
+	externalGatewayPodIPsAnnotations = "k8s.ovn.org/external-gw-podIPs"
+)
```
Per below let's have one location for this annotation definition, not two.

tssurya replied:
yep fixed this, moved it into util package.

dcbw commented on test/e2e/external_gateways.go:
```diff
@@ -984,6 +1153,27 @@ func pokeHostnameViaNC(podName, namespace, protocol, target string, port int) st
+	for _, pattern := range patterns {
+		for _, connEntry := range strings.Split(conntrackEntries, "\n") {
+			if strings.Contains(connEntry, pattern) && strings.Contains(connEntry, protocol) && strings.Contains(connEntry, podIP) {
+				klog.Infof("SURYA %v", connEntry)
```
@tssurya did you mean to leave this in there? :)

tssurya replied:
oopsy :D good eye detail haha, let me remove it.

tssurya commented on go-controller/pkg/ovn/egressgw.go:
```diff
@@ -139,6 +139,22 @@ func (oc *Controller) deleteRouteInfoLocked(name ktypes.NamespacedName) *externa
 	return routeInfo
 }
 
+func (oc *Controller) updateExternalGatewayPodsAnnotations(namespace string, exgwIPs []string) error {
```
oh right makes sense. didn't see the pattern there but yeah let me do a namespace_annotations.go

tssurya commented on go-controller/pkg/ovn/namespace.go:
```diff
@@ -89,9 +90,8 @@ func (oc *Controller) getRoutingPodGWs(nsInfo *namespaceInfo) map[string]gateway
-			gws:        make([]net.IP, len(v.gws)),
+			gws:        v.gws,
 		}
-		copy(item.gws, v.gws)
```
did the same here: `gws:        sets.NewString(v.gws.UnsortedList()...),`

dcbw commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,75 @@ func (n *OvnNode) WatchEndpoints() error {
+func exGatewayPodsAnnotationsChanged(oldNs, newNs *kapi.Namespace) bool {
+	// We only care about exgw pod deletions, i.e if the IPs on a namespace reduced than before, we return true
+	// NOTE: we don't care about the whole namespace getting deleted since pods will go away and so will the conntrack entries
+	return len(oldNs.Annotations[util.ExternalGatewayPodIPsAnnotation]) > len(newNs.Annotations[util.ExternalGatewayPodIPsAnnotation])
```
@tssurya this seems a bit fragile; it won't happen now due to the way we add/remove single IPs at a time, but I don't think we should rely on that. Should we instead do a strings.Split() and then compare lengths of the resulting arrays instead?

tssurya replied:
yup excellent point, what if its dual stack and IPV6 addresses lengths could mess up depending on how they look.. let me change this with what you suggested. thanks Dan as usual for your hawk eyes :)

dcbw replied:
@tssurya ideally we compare the actual contents too... but maybe that's not necessary right now given our single-ip-change behavior. But maybe we should just do the full compare now and be done with it... there won't be that many IPs so I don't think it would be very heavy.

tssurya replied:
hmm ok so thinking more about this, first step is to get the list of IPs from new annotation -> with the split done - and check if the "number of IPs" is less than before. but then you are right in the sense what if an IP got added and deleted at same time like with deployments where another one takes the place somehow and the length of my annotation stays the same....

I could do what you are suggesting, like really do a full blown comparision, would be as simple as old != new ? then trigger the arping etc.. but problem is we will do this for every SPK add as well, not just delete. Is that acceptable?

tssurya replied:
ACK let me change this logic in code per Dan's suggestions ^

dcbw commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,75 @@ func (n *OvnNode) WatchEndpoints() error {
+}
+
+func (n *OvnNode) checkAndDeleteStaleConntrackEntriesForNamespace(newNs *kapi.Namespace) {
+	validNextHopMAC := [][]byte{}
```
@tssurya can you change this to `validNextHopMACs` ?

tssurya commented:
/assign @trozet

trozet commented on go-controller/pkg/ovn/egressgw.go:
```diff
@@ -372,6 +379,9 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod *kapi.Pod, namespace str
 	// check if any gateways were stored for this pod
 	foundGws, ok := nsInfo.routingExternalPodGWs[podGWKey]
+	// update ns routing podIPs info
+	nsInfo.routingExternalGWs.gws.Delete(foundGws.gws.UnsortedList()...)
```
Looks like you are overloading nsInfo.routingExternalGWs.gws with pod ips now too? The comments in ovn.go weren't updated. Why is this necessary?

tssurya replied:
I see the confusion here, I probably should have put more comments...

There are two cases that can trigger this whole feature, 1) namespace annotation 2) SPK annotation. I am trying to reuse the variables here to then use that cache to fill the new podIPs annotation.

Let's say in method 1: namespace annotation, we already use the `nsInfo.routingExternalGWs` cache to store the gateway IPs that serve that namespace. So this was straight forward to then copy onto the new annotation.

In method 2: SPK annotation, we do the `routing-namespaces` annotation that determines which namespace this SPK is going to server. However we weren't updating that info onto the cache for those namespaces. So I reused the `existing cache and kept it updated for any given namespace what all external gateway IPs are serving it?` and then use that fill the new annotation.

Does that make sense?

tssurya replied:
@tssurya : no that doesn't make sense, don't mix the caches....

trozet commented on go-controller/pkg/node/node.go:
```diff
@@ -539,7 +539,16 @@ func (n *OvnNode) Start(ctx context.Context, wg *sync.WaitGroup) error {
+		if err != nil {
+			return fmt.Errorf("failed to watch namespaces: %w", err)
+		}
+		// every 5 minutes cleanup stale conntrack entries if any
```
5 minutes seems like too long. How come this value was chosen? ARP should be OK to try more often, maybe every minute?

tssurya replied:
it was a random interval :D I am open to suggestions... 1min is ok for me, TBH I have no idea on a large env what's a good value and what's not. I was thinking let's keep it at 5mins and user decides this is too infrequent we could fix it later, but I didn't want the reverse of too many ARPs.

tssurya replied:
Keeping it at 3mins compromise for now, because the ARP timeout is 50 milliseconds, that means at least we will have time for 60 ARP timeouts within the 3mins, if its 1 min it reduces to 20 ARP timeouts at max worst case scenario.

trozet replied:
i think this is more reason we need to parallelize the ARPs, we dont want to be waiting that long. Default ARP timeout in linux is 30 seconds. I dont see a problem with running every minute.

tssurya commented:
/retest-failed

trozet commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,79 @@ func (n *OvnNode) WatchEndpoints() error {
+	if err != nil {
+		klog.Errorf("Unable to get pods from informer: %v", err)
+	}
+	for _, namespace := range namespaces {
```
so this will iterate all namespaces, and then in each namespace, serially arp for each GW IP. Would it be better to parallelize at least the ARPs? Maybe it's overkill...

tssurya replied:
probably yes I was under the impression not many namespaces will have the exgw related annotations, but I could be wrong.

tssurya replied:
done parallelized this into routines within loop

tssurya commented on test/e2e/external_gateways.go:
```diff
@@ -374,6 +377,108 @@ var _ = ginkgo.Describe("e2e multiple external gateway validation", func() {
+		ginkgo.By("Check if conntrack entries for ECMP routes are removed for the deleted external gateway if traffic is UDP")
+		res = pokeConntrackEntries(nodeName, addresses.srcPodIP, protocol, macAddressGW)
+		if protocol == "udp" {
+			gomega.Expect(res).To(gomega.Equal(1))
```
because we added 2 SPKs (gwContainer1, gwContainer2) for this pod and then we deleted one SPK pod (gwContainer1), so we still have the conntrack entry for the second SPK pod (gwContainer2).

tssurya replied:
added a comment to make this clear.

tssurya commented:
@trozet : I think I addressed all your comments PTAL, 2 things 

1) should I parallelize the ARP by doing them in routines within the loop? If yes then I guess I need a lock on the final MAC result variable and append whenever I get the response back? 
2) are you ok with a 3min interval for the periodic cleanup as a compromise? :D If 5 is too less its possible 1 is too much? I have asked Nabeel for input regarding approximate number of MACs we might have to deal with.

trozet commented:
> @trozet : I think I addressed all your comments PTAL, 2 things
> 
> 1. should I parallelize the ARP by doing them in routines within the loop? If yes then I guess I need a lock on the final MAC result variable and append whenever I get the response back?
> 2. are you ok with a 3min interval for the periodic cleanup as a compromise? :D If 5 is too less its possible 1 is too much? I have asked Nabeel for input regarding approximate number of MACs we might have to deal with.

You can just use a sync map to store all of the found macs, then you dont need to worry about locking.

tssurya commented on go-controller/pkg/node/node.go:
```diff
@@ -539,7 +539,16 @@ func (n *OvnNode) Start(ctx context.Context, wg *sync.WaitGroup) error {
+		if err != nil {
+			return fmt.Errorf("failed to watch namespaces: %w", err)
+		}
+		// every 3 minutes cleanup stale conntrack entries if any
```
done.

tssurya commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,82 @@ func (n *OvnNode) WatchEndpoints() error {
+		UpdateFunc: func(old, new interface{}) {
+			oldNs, newNs := old.(*kapi.Namespace), new.(*kapi.Namespace)
+			if exGatewayPodsAnnotationsChanged(oldNs, newNs) {
+				n.checkAndDeleteStaleConntrackEntriesForNamespace(newNs)
```
ahh I see your point.. hmm let me see if I can open an issue to fix this up in future.

tssurya commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,95 @@ func (n *OvnNode) WatchEndpoints() error {
+		klog.Errorf("Unable to get pods from informer: %v", err)
+	}
+	for _, namespace := range namespaces {
+		if (len(namespace.Annotations[util.ExternalGatewayPodIPsAnnotation]) > 0) || (len(namespace.Annotations[util.RoutingExternalGWsAnnotation]) > 0) {
```
@trozet had another nice idea for improvement. If this node has no local pods on this namespace then we can exit straight away.

tssurya replied:
looking good!
```
$ oc logs -n ovn-kubernetes ovnkube-node-xgrvb -c ovnkube-node | grep SURYA
I0712 19:13:15.011095     881 node.go:678] SURYA [10.244.0.3]
I0712 19:13:49.923939     881 node.go:678] SURYA [10.244.0.3]
I0712 19:14:49.983622     881 node.go:678] SURYA [10.244.0.3]
I0712 19:15:50.054568     881 node.go:678] SURYA [10.244.0.3]
I0712 19:16:50.123751     881 node.go:678] SURYA [10.244.0.3]
$ oc logs -n ovn-kubernetes ovnkube-node-fbt5g -c ovnkube-node | grep SURYA
$ oc logs -n ovn-kubernetes ovnkube-node-wn8xv -c ovnkube-node | grep SURYA
I0712 18:39:28.765682    1883 node.go:678] SURYA [10.244.2.6]
I0712 18:39:36.748368    1883 node.go:678] SURYA [10.244.2.6]
[surya@hidden-temple e2e]$ 

```

trozet reviewed (approved):
thanks @tssurya !

dcbw commented on go-controller/pkg/node/node.go:
```diff
@@ -595,6 +604,103 @@ func (n *OvnNode) WatchEndpoints() error {
+		klog.Errorf("Unable to get pods from informer: %v", err)
+	}
+	for _, namespace := range namespaces {
+		if (len(namespace.Annotations[util.ExternalGatewayPodIPsAnnotation]) > 0) || (len(namespace.Annotations[util.RoutingExternalGWsAnnotation]) > 0) {
```
@tssurya hmm; do we need this code to trigger when all gateways are removed from the namespace? Or is that handled somewhere else?

tssurya replied:
when all gateways are removed from namespace we would hit the oldNs != newNs below for the event triggered mechanism right? after that we don't need any periodic ARPs for empty annotations right?

tssurya replied:
however I could just change this to "if namespace is annotated" if that makes it better

tssurya replied:
actually you have a point. I don't think we ever do a allowMACList={empty} pass to deletion filter. :/

tssurya commented:
Making it easy for reviewers:
What changed in the new push:

e2e test changed to include empty list annotation case and now it probes total conntrack entries for a pod and ensures we do not accidentally nuke all the pod's conntrack entries if labels=nil (which it won't be from the new push)
```
@@ -1233,12 +1279,16 @@ func pokeConntrackEntries(nodeName, podIP, protocol string, patterns []string) i
        conntrackEntries, err := framework.RunKubectl("ovn-kubernetes", args...)
        framework.ExpectNoError(err, "failed to get the conntrack entries from node %s", nodeName)
        numOfConnEntries := 0
-       for _, pattern := range patterns {
-               for _, connEntry := range strings.Split(conntrackEntries, "\n") {
-                       if strings.Contains(connEntry, pattern) && strings.Contains(connEntry, protocol) && strings.Contains(connEntry, podIP) {
+       for _, connEntry := range strings.Split(conntrackEntries, "\n") {
+               match := strings.Contains(connEntry, protocol) && strings.Contains(connEntry, podIP)
+               for _, pattern := range patterns {
+                       if strings.Contains(connEntry, pattern) && match {
                                numOfConnEntries++
                        }
                }
+               if len(patterns) == 0 && match {
+                       numOfConnEntries++
+               }
        }
```
AND main logic around when to trigger a stale check: cc @dcbw 
```
        for _, namespace := range namespaces {                                                                                                                               
-               if (len(namespace.Annotations[util.ExternalGatewayPodIPsAnnotation]) > 0) || (len(namespace.Annotations[util.RoutingExternalGWsAnnotation]) > 0) {           
+               _, foundRoutingExternalGWsAnnotation := namespace.Annotations[util.RoutingExternalGWsAnnotation]                                                             
+               _, foundExternalGatewayPodIPsAnnotation := namespace.Annotations[util.ExternalGatewayPodIPsAnnotation]                                                       
+               if foundRoutingExternalGWsAnnotation || foundExternalGatewayPodIPsAnnotation {                                                                               
                        pods, err := n.watchFactory.GetPods(namespace.Name)                                                                                                  
                        if err != nil {                                                                                                                                      
                                klog.Warningf("Unable to get pods from informer for namespace %s: %v", namespace.Name, err)                                                  
@@ -667,6 +669,12 @@ func (n *OvnNode) checkAndDeleteStaleConntrackEntriesForNamespace(newNs *kapi.Na                                                                        
                validNextHopMACs = append(validNextHopMACs, value.([]byte))                                                                                                  
                return true                                                                                                                                                  
        })                                                                                                                                                                   
+       // Handle corner case where there are 0 IPs on the annotations OR none of the ARPs were successful; i.e allowMACList={empty}.                                        
+       // This means we *need to* pass a label > 128 bits that will not match on any conntrack entry labels for these pods.                                                 
+       // That way any remaining entries with labels having MACs set will get purged.                                                                                       
+       if len(validNextHopMACs) == 0 {                                                                                                                                      
+               validNextHopMACs = append(validNextHopMACs, []byte("does-not-contain-anything"))                                                                             
+       } 
```
